### ISSUE TITLE
[Patterns] How should elements be added to a distributed list?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#18967]
This issue is part of a series of issues to design the interface for collections across serial, parallel, and distributed contexts. Our goal is to determine what we think is reasonable to support and what doesn't seem useful. Additional patterns welcome (but it would be best to put them in their own issue for discussion, likely)

Would it make sense to provide a block or cyclic "distribution" for lists? It seems like cyclic would be the easiest to support in a load balanced manner, while maybe blockcyclic would make more sense than block since the end size of the list won't likely be known ahead of time (otherwise maybe the user would use an array)

```chapel
var l1 = new DistributedList(...);

for loc in Locales {
  on loc { // Should this have an impact on where the data will ultimately live?
    for x in 0..19 {
      l1.append(x); // Note: should append be callable in parallel?  It won't lead to a deterministic ordering . . .
    }
  }
}
```

**Comment by mppf** [comment]
>   on loc { // Should this have an impact on where the data will ultimately live?

I don't think so, but of course one can create a class keep the instance stored where you want that way.

>       l1.append(x); // Note: should append be callable in parallel?  It won't lead to a deterministic ordering . . .

I think it should be callable in parallel.

> Would it make sense to provide a block or cyclic "distribution" for lists? It seems like cyclic would be the easiest to support in a load balanced manner, while maybe blockcyclic would make more sense than block since the end size of the list won't likely be known ahead of time (otherwise maybe the user would use an array)

IMO it would be reasonable to start with just BlockCyclic with a configurable block size. That's fairly general. But presumably we would want to have a use case in mind before implementing something like this, right?

**Comment by vasslitvinov** [comment]
On the topic of parallelism, if two tasks add elements concurrently, should the globally-visible order of elements be interlaced or batched? The answer may be different if the tasks synchronize with each other vs. don't.

If elements are added using a forall, the global order needs to be settled at the point when the forall loop completes. I can see an optimization that batches up the elements added by each task, analogously to what we did for the global array updates in RA [right?]. There is of course room for using an explicit aggregator, which makes semantic questions simpler.
